design/namespaced-roles.md

```diff
@@ -0,0 +1,349 @@
+`role_v1:/ud/groups/devops/developer`
+`role_v1:/ud/groups/devops/devops_role`
+
+Now these roles are assigned directly to `UserA` and `UserB` so that:
+
+- `UserA` has the `role_v1:/ud/groups/iam/manager` and the `role_v1:/ud/groups/devops/developer` roles.
+- `UserB` has the `role_v1:/ud/groups/iam/developer` and the `role_v1:/ud/groups/devops/developer` roles.
+
+Also, we should be able to still support giving everyone in a group an inherited role, such as:
+
+- `Group:devops` has the `role_v1:/ud/groups/devops/devops_role` role or `role_v1:/ud/devops_role`.
+
+#### Representation in the tokens
+
+So now that we have a namespaced role attached to a user, a **possible representation** in the tokens, defined by a token mappers could look like:
+
+- For UserA
+
+```json
+{
+    ....
+    "roles": {
+        "groups": {
+            "iam": ["manager"],
+            "devops": ["developer", "devops_role"]
+        }
+    }
+    ....
+}
+
+```
+
+As mappers can be custom implementations, users would be able to define other representations, or Keycloak could provide more built-in mappers to do so. 
+
+An alternative representation could look like:
+
+```json
+{
+    ....
+    "roles": [
+        "groups/iam/manager",
+        "groups/devops/developer",
+        "groups/devops/devops_role"
+    ]
+    ....
+}
+
+```
+
+- For UserB
+
+```json
+{
+    ....
+    "roles": {
+        "groups": {
+            "iam": ["developer"],
+            "devops": ["developer", "devops_role"]
+        }
+    }
+    ....
+}
+```
+
+The whole idea is that by leveraging namespaces, users can define the structure that they need for their business needs.
+
+### Use Case 2 - Isolate roles to specific entities
+
+It will also be possible to isole certain roles to specific entities such as tenants.
+
+- `role_v1:/ud/tenants/tenant1/developer` would only be assignable in `tenant1` while
+- `role_v1:/ud/tenants/tenant2/somethingelse` would only be assignable in `tenant2`.
+
+Or just to specific clients within the realm:
+
+- `role_v1:/kc/clients/client-a/client-admin` would only give users the `client-admin` role when interacting with the `client-a` client.
+- `role_v1:/kc/groups/devops/group-admin` would only give users the `group-admin` role when interacting with with the `devops` groups.
+
+### Use Case 3 - Custom, resource server specific roles
+
+Another interesting use case is the creation of roles that are not directly restricted to Keycloak's entities such as Realms, Groups, Clients etc.
+
+Right now in Keycloak, clients are effectively a representation of a Resource Server (among other things). You can assign roles to clients and those roles would only be usable on those clients.
+
+As previously mentioned, a role would be namespaced to a certain client such as:
+
+-  `role_v1:/ud/clients/{client-name}/{role-name}`
+
+or with a regexp, they could be namespaced to a specific subset of clients:
+
+- `role_v1:/ud/clients/{regexp-containing-wildcard}/{role-name}`
```

Comment:
Wouldn't this be incredibly complex to resolve? Also, we'd want to use client-id to specify a client rather than client-name, and often these are just generated (and really should be), so is there a meaningful regex here?